File: backend/app/api/profiles.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.models.profile import Profile
from app.models.user import User
from app.models.auto_reply import AutoReply
from app.models.text_example import TextExample
from app.services.twilio_service import TwilioService
from app.extensions import db
import json

profiles_bp = Blueprint('profiles', __name__)

@profiles_bp.route('', methods=['GET'])
@jwt_required()
def get_profiles():
    user_id = get_jwt_identity()
    
    
    # Fetch profiles
    profiles = Profile.query.filter_by(user_id=user_id).all()
    
    return jsonify([profile.to_dict() for profile in profiles]), 200

@profiles_bp.route('/<int:profile_id>', methods=['GET'])
@jwt_required()
def get_profile(profile_id):
    user_id = get_jwt_identity()
    
    # Find profile
    profile = Profile.query.get_or_404(profile_id)
    
    # Check ownership
    if profile.user_id != user_id:
        return jsonify({"error": "Unauthorized"}), 403
    
    return jsonify(profile.to_dict()), 200

@profiles_bp.route('', methods=['POST'])
@jwt_required()
def create_profile():
    user_id = get_jwt_identity()
    user = User.query.get(user_id)
    data = request.json
    
    # Validate required fields
    if not all([data.get('name')]):
        return jsonify({"error": "Profile name is required"}), 400
    
    # Check if user has a Twilio account
    if not user.twilio_account_sid:
        return jsonify({
            "error": "You need to set up a Twilio account first",
            "details": "Please set up a Twilio account in your account settings before creating a profile"
        }), 400
    
    # Determine phone number acquisition method
    phone_number = None
    phone_source = data.get('phone_source', 'new')  # 'new', 'existing', 'manual'
    
    if phone_source == 'new':
        # Purchase a new number
        area_code = data.get('area_code')
        country_code = data.get('country_code', 'US')
        
        # Create TwilioService instance with the user
        twilio_service = TwilioService(user)
        success, result = twilio_service.purchase_phone_number(area_code, country_code)
        
        if not success:
            return jsonify({"error": f"Failed to purchase phone number: {result}"}), 400
        
        phone_number = result
        
    elif phone_source == 'existing':
        # Use an existing number from the user's Twilio account
        phone_number = data.get('phone_number')
        if not phone_number:
            return jsonify({"error": "Phone number is required when using an existing number"}), 400
        
        # Create TwilioService instance with the user
        twilio_service = TwilioService(user)
        success, result = twilio_service.setup_sms_webhook(phone_number)
        
        if not success:
            return jsonify({"error": f"Failed to configure webhook for phone number: {result}"}), 400
        
    elif phone_source == 'manual':
        # Manually entered number (no Twilio provisioning)
        phone_number = data.get('phone_number')
        if not phone_number:
            return jsonify({"error": "Phone number is required when manually entering a number"}), 400
    
    else:
        return jsonify({"error": "Invalid phone_source value"}), 400
    
    # Create default business hours
    default_hours = {
        "monday": {"start": "10:00", "end": "22:00"},
        "tuesday": {"start": "10:00", "end": "22:00"},
        "wednesday": {"start": "10:00", "end": "22:00"},
        "thursday": {"start": "10:00", "end": "22:00"},
        "friday": {"start": "10:00", "end": "22:00"},
        "saturday": {"start": "12:00", "end": "22:00"},
        "sunday": {"start": "12:00", "end": "22:00"},
    }
    
    # Create new profile
    profile = Profile(
        user_id=user_id,
        name=data['name'],
        phone_number=phone_number,
        description=data.get('description', ''),
        business_hours=json.dumps(default_hours)
    )
    
    db.session.add(profile)
    db.session.commit()
    
    return jsonify(profile.to_dict()), 201

@profiles_bp.route('/<int:profile_id>/toggle_ai', methods=['POST'])
@jwt_required()
def toggle_ai(profile_id):
    user_id = get_jwt_identity()
    data = request.json
    
    # Find profile
    profile = Profile.query.get_or_404(profile_id)
    
    # Check ownership
    if profile.user_id != user_id:
        return jsonify({"error": "Unauthorized"}), 403
    
    # Update AI setting
    profile.ai_enabled = data.get('enabled', not profile.ai_enabled)
    db.session.commit()
    
    return jsonify({
        "success": True,
        "ai_enabled": profile.ai_enabled
    }), 200

# Auto-reply endpoints
@profiles_bp.route('/<int:profile_id>/auto_replies', methods=['GET'])
@jwt_required()
def get_auto_replies(profile_id):
    user_id = get_jwt_identity()
    
    # Find profile
    profile = Profile.query.get_or_404(profile_id)
    
    # Check ownership
    if profile.user_id != user_id:
        return jsonify({"error": "Unauthorized"}), 403
    
    # Get auto replies
    auto_replies = AutoReply.query.filter_by(profile_id=profile_id).all()
    
    return jsonify([ar.to_dict() for ar in auto_replies]), 200

# Text examples endpoints
@profiles_bp.route('/<int:profile_id>/text_examples', methods=['GET'])
@jwt_required()
def get_text_examples(profile_id):
    user_id = get_jwt_identity()
    
    # Find profile
    profile = Profile.query.get_or_404(profile_id)
    
    # Check ownership
    if profile.user_id != user_id:
        return jsonify({"error": "Unauthorized"}), 403
    
    # Get text examples
    examples = TextExample.query.filter_by(profile_id=profile_id).all()
    
    return jsonify([ex.to_dict() for ex in examples]), 200

File: frontend/src/api/profiles.ts
import axios from './index';
import type { Profile, AutoReply, TextExample, AIModelSettings } from '../types';

// Get all profiles for current user
export const getProfiles = async () => {
  const response = await axios.get('/api/profiles');
  return response.data as Profile[];
};

// Get a specific profile
export const getProfile = async (profileId: number) => {
  const response = await axios.get(`/api/profiles/${profileId}`);
  return response.data as Profile;
};

// Create a new profile
export const createProfile = async (profileData: Partial<Profile>) => {
  const response = await axios.post('/api/profiles', profileData);
  return response.data as Profile;
};

// Update a profile
export const updateProfile = async (profileId: number, profileData: Partial<Profile>) => {
  const response = await axios.put(`/api/profiles/${profileId}`, profileData);
  return response.data as Profile;
};

// Delete a profile
export const deleteProfile = async (profileId: number) => {
  const response = await axios.delete(`/api/profiles/${profileId}`);
  return response.data;
};

// Toggle AI for a profile
export const toggleAI = async (profileId: number, enabled: boolean) => {
  const response = await axios.post(`/api/profiles/${profileId}/toggle_ai`, { enabled });
  return response.data;
};

// Get auto replies for a profile
export const getAutoReplies = async (profileId: number) => {
  const response = await axios.get(`/api/profiles/${profileId}/auto_replies`);
  return response.data as AutoReply[];
};

// Create auto reply
export const createAutoReply = async (profileId: number, data: { keyword: string; response: string; priority?: number }) => {
  const response = await axios.post(`/api/profiles/${profileId}/auto_replies`, data);
  return response.data as AutoReply;
};

// Update auto reply
export const updateAutoReply = async (profileId: number, replyId: number, data: Partial<AutoReply>) => {
  const response = await axios.put(`/api/profiles/${profileId}/auto_replies/${replyId}`, data);
  return response.data as AutoReply;
};

// Delete auto reply
export const deleteAutoReply = async (profileId: number, replyId: number) => {
  const response = await axios.delete(`/api/profiles/${profileId}/auto_replies/${replyId}`);
  return response.data;
};

// Get text examples for a profile
export const getTextExamples = async (profileId: number) => {
  const response = await axios.get(`/api/profiles/${profileId}/text_examples`);
  return response.data as TextExample[];
};

// Add text example
export const addTextExample = async (profileId: number, data: { content: string; is_incoming: boolean }) => {
  const response = await axios.post(`/api/profiles/${profileId}/text_examples`, data);
  return response.data as TextExample;
};

// Delete text example
export const deleteTextExample = async (profileId: number, exampleId: number) => {
  const response = await axios.delete(`/api/profiles/${profileId}/text_examples/${exampleId}`);
  return response.data;
};

// Get AI model settings
export const getAISettings = async (profileId: number) => {
  const response = await axios.get(`/api/profiles/${profileId}/ai_settings`);
  return response.data as AIModelSettings;
};

// Update AI model settings
export const updateAISettings = async (profileId: number, data: Partial<AIModelSettings>) => {
  const response = await axios.put(`/api/profiles/${profileId}/ai_settings`, data);
  return response.data as AIModelSettings;
};

// Set business hours
export const setBusinessHours = async (profileId: number, businessHours: any) => {
  const response = await axios.post(`/api/profiles/${profileId}/business_hours`, { business_hours: businessHours });
  return response.data;
};

// Get out of office reply
export const getOutOfOfficeReply = async (profileId: number) => {
  const response = await axios.get(`/api/profiles/${profileId}/out_of_office`);
  return response.data;
};

// Update out of office reply
export const updateOutOfOfficeReply = async (profileId: number, message: string, isActive: boolean) => {
  const response = await axios.put(`/api/profiles/${profileId}/out_of_office`, { message, is_active: isActive });
  return response.data;
};